chainerx_cc/chainerx/native/native_device/dot.cc

```diff
@@ -166,16 +175,32 @@ void NativeDevice::Dot(const Array& a, const Array& b, const Array& out) {
         CHAINERX_ASSERT(out.shape()[0] == m);
         CHAINERX_ASSERT(out.shape()[1] == n);
 
+        using AccT = typename std::conditional<std::is_same<T, Float16>{}, float, T>::type;
+        constexpr auto acc_dtype = PrimitiveType<AccT>::kDtype;
+
+        Array acc = out.AsType(acc_dtype, false);
+        acc.Fill(0);
```

Comment:
A nitpick but this can be done in the original place (outside `VisitDtype`).
It saves a bit of code size.